src/jni/src/datamanlite/RowReader.h

```diff
@@ -0,0 +1,283 @@
+        Iterator(Iterator&& iter);
+        const Cell& operator*() const;
+        const Cell* operator->() const;
+        Iterator& operator++();
+        operator bool() const;
+        bool operator==(const Iterator& rhs) const noexcept;
+    private:
+        const RowReader* reader_;
+        const size_t numFields_;
+        std::unique_ptr<Cell> cell_;
+        int64_t index_ = 0;
+        int32_t bytes_ = 0;
+        int64_t offset_ = 0;
+
+        Iterator(const RowReader* reader, size_t numFields, int64_t index = 0);
+    };
+
+
+public:
+    static std::unique_ptr<RowReader> getRowReader(const Slice& row,
+                                                   std::shared_ptr<const SchemaProviderIf> schema);
+
+    virtual ~RowReader() = default;
+
+    SchemaVer schemaVer() const noexcept;
+    int32_t numFields() const noexcept;
+
+    Iterator begin() const noexcept;
+    Iterator end() const noexcept;
+
+    ResultType getBool(const Slice& name, bool& v) const noexcept;
+    ResultType getBool(int64_t index, bool& v) const noexcept;
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value, ResultType>::type
+    getInt(const Slice& name, T& v) const noexcept;
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value, ResultType>::type
+    getInt(int64_t index, T& v) const noexcept;
+
+    ResultType getFloat(const Slice& name, float& v) const noexcept;
+    ResultType getFloat(int64_t index, float& v) const noexcept;
+
+    ResultType getDouble(const Slice& name, double& v) const noexcept;
+    ResultType getDouble(int64_t index, double& v) const noexcept;
+
+    ResultType getString(const Slice& name, Slice& v) const noexcept;
+    ResultType getString(int64_t index, Slice& v) const noexcept;
+
+    ResultType getVid(const Slice& name, int64_t& v) const noexcept;
+    ResultType getVid(int64_t index, int64_t& v) const noexcept;
+
+
+    std::shared_ptr<const SchemaProviderIf> getSchema() const {
+        return schema_;
+    }
+
+    // TODO getPath(const std::string& name) const noexcept;
+    // TODO getPath(int64_t index) const noexcept;
+    // TODO getList(const std::string& name) const noexcept;
+    // TODO getList(int64_t index) const noexcept;
+    // TODO getSet(const std::string& name) const noexcept;
+    // TODO getSet(int64_t index) const noexcept;
+    // TODO getMap(const std::string& name) const noexcept;
+    // TODO getMap(int64_t index) const noexcept;
+
+private:
+    std::shared_ptr<const SchemaProviderIf> schema_;
+
+    Slice data_;
+    int32_t headerLen_ = 0;
+    int32_t numBytesForOffset_ = 0;
+    // Block offet value is composed by two integers. The first one is
+    // the block offset, the second one is the largest index being visited
+    // in the block. This index is zero-based
+    mutable std::vector<std::pair<int64_t, uint8_t>> blockOffsets_;
+    mutable std::vector<int64_t> offsets_;
+
+private:
+    static int32_t getSchemaVer(Slice row);
```

Comment:
Do we need to unify the type of schema version? The `SchemaVer` used in thrift is int64_t

Reply:
Good point